In [ ]:
import tensorflow as tf
print(tf.__version__)


2.15.0


In [ ]:
from tensorflow.keras.datasets import cifar10
from tensorflow.keras.utils import to_categorical

# Įkelti CIFAR-10 duomenų rinkinį
(x_train, y_train), (x_test, y_test) = cifar10.load_data()

# Normalizuoti duomenis į intervalą [0, 1]
x_train = x_train.astype('float32') / 255.0
x_test = x_test.astype('float32') / 255.0

# Kategoriniai etiketės
y_train = to_categorical(y_train, 10)
y_test = to_categorical(y_test, 10)

# Patikrinti duomenų formą
print(f"x_train shape: {x_train.shape}, y_train shape: {y_train.shape}")
print(f"x_test shape: {x_test.shape}, y_test shape: {y_test.shape}")


170498071/170498071 [==============================] - 5s 0us/step
x_train shape: (50000, 32, 32, 3), y_train shape: (50000, 10)
x_test shape: (10000, 32, 32, 3), y_test shape: (10000, 10)


In [ ]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, Dropout

def build_cnn_model():
    model = Sequential()
    model.add(Conv2D(32, (3, 3), activation='relu', input_shape=(32, 32, 3)))
    model.add(MaxPooling2D((2, 2)))
    model.add(Conv2D(64, (3, 3), activation='relu'))
    model.add(MaxPooling2D((2, 2)))
    model.add(Conv2D(128, (3, 3), activation='relu'))
    model.add(Flatten())
    model.add(Dense(128, activation='relu'))
    model.add(Dropout(0.5))
    model.add(Dense(10, activation='softmax'))
    return model

model = build_cnn_model()
model.summary()


Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d (Conv2D)             (None, 30, 30, 32)        896       
                                                                 
 max_pooling2d (MaxPooling2  (None, 15, 15, 32)        0         
 D)                                                              
                                                                 
 conv2d_1 (Conv2D)           (None, 13, 13, 64)        18496     
                                                                 
 max_pooling2d_1 (MaxPoolin  (None, 6, 6, 64)          0         
 g2D)                                                            
                                                                 
 conv2d_2 (Conv2D)           (None, 4, 4, 128)         73856     
                                                                 
 flatten (Flatten)           (None, 2048)              0

In [ ]:
# Modelio kompiliacija
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

# Modelio treniravimas
history = model.fit(x_train, y_train, epochs=10, batch_size=64, validation_data=(x_test, y_test))


Epoch 1/10
782/782 [==============================] - 74s 93ms/step - loss: 1.6435 - accuracy: 0.3930 - val_loss: 1.3497 - val_accuracy: 0.4957
Epoch 2/10
782/782 [==============================] - 70s 89ms/step - loss: 1.2937 - accuracy: 0.5384 - val_loss: 1.0882 - val_accuracy: 0.6186
Epoch 3/10
782/782 [==============================] - 69s 89ms/step - loss: 1.1404 - accuracy: 0.6002 - val_loss: 1.0026 - val_accuracy: 0.6427
Epoch 4/10
782/782 [==============================] - 73s 93ms/step - loss: 1.0353 - accuracy: 0.6390 - val_loss: 0.9481 - val_accuracy: 0.6702
Epoch 5/10
782/782 [==============================] - 73s 94ms/step - loss: 0.9507 - accuracy: 0.6669 - val_loss: 1.0082 - val_accuracy: 0.6490
Epoch 6/10
782/782 [==============================] - 70s 90ms/step - loss: 0.8703 - accuracy: 0.6960 - val_loss: 0.8614 - val_accuracy: 0.7009
Epoch 7/10
782/782 [==============================] - 70s 89ms/step - loss: 0.8240 - accuracy: 0.7118 - val_loss: 0.8554 - val_accuracy:

In [ ]:
# Modelio įvertinimas
loss, accuracy = model.evaluate(x_test, y_test)
print(f'Test accuracy: {accuracy}')


313/313 [==============================] - 6s 18ms/step - loss: 0.7930 - accuracy: 0.7267
Test accuracy: 0.7267000079154968


In [ ]:
# Padidinkime epochų skaičių iki 20
history = model.fit(x_train, y_train, epochs=20, batch_size=64, validation_data=(x_test, y_test))

# Modelio įvertinimas
loss, accuracy = model.evaluate(x_test, y_test)
print(f'Test accuracy after increasing epochs to 20: {accuracy}')


Epoch 1/20
782/782 [==============================] - 71s 90ms/step - loss: 0.6487 - accuracy: 0.7740 - val_loss: 0.7864 - val_accuracy: 0.7339
Epoch 2/20
782/782 [==============================] - 70s 89ms/step - loss: 0.6215 - accuracy: 0.7817 - val_loss: 0.8127 - val_accuracy: 0.7330
Epoch 3/20
782/782 [==============================] - 70s 89ms/step - loss: 0.5909 - accuracy: 0.7923 - val_loss: 0.7873 - val_accuracy: 0.7394
Epoch 4/20
782/782 [==============================] - 72s 93ms/step - loss: 0.5569 - accuracy: 0.8013 - val_loss: 0.8613 - val_accuracy: 0.7316
Epoch 5/20
782/782 [==============================] - 71s 91ms/step - loss: 0.5264 - accuracy: 0.8131 - val_loss: 0.8182 - val_accuracy: 0.7375
Epoch 6/20
782/782 [==============================] - 71s 90ms/step - loss: 0.5059 - accuracy: 0.8212 - val_loss: 0.8504 - val_accuracy: 0.7378
Epoch 7/20
782/782 [==============================] - 70s 90ms/step - loss: 0.4818 - accuracy: 0.8273 - val_loss: 0.9217 - val_accuracy:

In [ ]:
# Pakeiskime partijos dydį į 128
history = model.fit(x_train, y_train, epochs=10, batch_size=128, validation_data=(x_test, y_test))

# Modelio įvertinimas
loss, accuracy = model.evaluate(x_test, y_test)
print(f'Test accuracy after changing batch size to 128: {accuracy}')


Epoch 1/10
391/391 [==============================] - 70s 179ms/step - loss: 0.2203 - accuracy: 0.9188 - val_loss: 1.1897 - val_accuracy: 0.7415
Epoch 2/10
391/391 [==============================] - 69s 176ms/step - loss: 0.1898 - accuracy: 0.9278 - val_loss: 1.2866 - val_accuracy: 0.7407
Epoch 3/10
391/391 [==============================] - 68s 174ms/step - loss: 0.1890 - accuracy: 0.9302 - val_loss: 1.2918 - val_accuracy: 0.7401
Epoch 4/10
391/391 [==============================] - 68s 173ms/step - loss: 0.1912 - accuracy: 0.9284 - val_loss: 1.3628 - val_accuracy: 0.7387
Epoch 5/10
391/391 [==============================] - 66s 170ms/step - loss: 0.1820 - accuracy: 0.9324 - val_loss: 1.3754 - val_accuracy: 0.7415
Epoch 6/10
391/391 [==============================] - 66s 169ms/step - loss: 0.1787 - accuracy: 0.9328 - val_loss: 1.3807 - val_accuracy: 0.7389
Epoch 7/10
391/391 [==============================] - 66s 170ms/step - loss: 0.1808 - accuracy: 0.9318 - val_loss: 1.4176 - val_ac

In [ ]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, Dropout

def build_cnn_model_v2():
    model = Sequential()
    model.add(Conv2D(32, (3, 3), activation='relu', input_shape=(32, 32, 3)))
    model.add(MaxPooling2D((2, 2)))
    model.add(Conv2D(64, (3, 3), activation='relu'))
    model.add(MaxPooling2D((2, 2)))
    model.add(Conv2D(128, (3, 3), activation='relu'))
    model.add(MaxPooling2D((2, 2)))
    model.add(Flatten())
    model.add(Dense(128, activation='relu'))
    model.add(Dropout(0.5))
    model.add(Dense(10, activation='softmax'))
    return model

model_v2 = build_cnn_model_v2()
model_v2.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

# Modelio treniravimas su nauju sluoksniu
history = model_v2.fit(x_train, y_train, epochs=10, batch_size=64, validation_data=(x_test, y_test))

# Modelio įvertinimas
loss, accuracy = model_v2.evaluate(x_test, y_test)
print(f'Test accuracy after adding an additional convolutional layer: {accuracy}')


Epoch 1/10
782/782 [==============================] - 76s 96ms/step - loss: 1.6997 - accuracy: 0.3714 - val_loss: 1.3422 - val_accuracy: 0.5176
Epoch 2/10
782/782 [==============================] - 70s 90ms/step - loss: 1.3581 - accuracy: 0.5099 - val_loss: 1.2253 - val_accuracy: 0.5529
Epoch 3/10
782/782 [==============================] - 68s 87ms/step - loss: 1.2020 - accuracy: 0.5741 - val_loss: 1.1285 - val_accuracy: 0.6040
Epoch 4/10
782/782 [==============================] - 68s 87ms/step - loss: 1.0984 - accuracy: 0.6159 - val_loss: 1.0125 - val_accuracy: 0.6477
Epoch 5/10
782/782 [==============================] - 68s 87ms/step - loss: 1.0048 - accuracy: 0.6505 - val_loss: 0.9741 - val_accuracy: 0.6578
Epoch 6/10
782/782 [==============================] - 68s 87ms/step - loss: 0.9361 - accuracy: 0.6728 - val_loss: 0.9112 - val_accuracy: 0.6836
Epoch 7/10
782/782 [==============================] - 69s 89ms/step - loss: 0.8769 - accuracy: 0.6940 - val_loss: 0.9075 - val_accuracy:

# CIFAR-10 Vaizdų Klasifikacija Naudojant CNN

Šio projekto tikslas yra sukurti konvoliucinį neuroninį tinklą (CNN) CIFAR-10 vaizdų rinkiniui klasifikuoti. CNN modelis naudoja kelių sluoksnių architektūrą, kad atpažintų ir klasifikuotų vaizdus į 10 kategorijų.



## Duomenų Rinkinio Aprašymas

CIFAR-10 yra populiarus duomenų rinkinys, kurį sudaro 60000 32x32 spalvotų vaizdų, suskirstytų į 10 klasių, kiekviena klasė turi 6000 vaizdų.


## Modelio Architektūra

CNN modelis susideda iš trijų konvoliucinių sluoksnių su ReLU aktyvacijos funkcijomis, po kurių seka maksimalaus susiliejimo (MaxPooling) sluoksniai. Po konvoliucinių sluoksnių eina tankus sluoksnis su Dropout reguliavimu ir galiausiai išėjimo sluoksnis su Softmax aktyvacija.


## Treniravimo Procesas ir Eksperimentai

Modelis buvo treniruojamas naudojant Adam optimizatorių ir kategorinį kryžminį entropijos nuostolį. Treniruotės metu buvo naudojama 10 epochų su 64 partijos dydžiu.


### Eksperimentas 1: Epochų skaičiaus padidinimas iki 20

Tikslumas po 20 epochų - 72.94%


### Eksperimentas 2: Partijos dydžio keitimas į 128

Tikslumas su 128 partijos dydžiu - 73.23%


### Eksperimentas 3: Papildomo konvoliucinio sluoksnio pridėjimas

Tikslumas su papildomu sluoksniu - 70.90%


## Modelio Įvertinimas ir Išvados

Modelis buvo įvertintas testavimo duomenų rinkinyje, kur pasiekė 72.67% tikslumą. Eksperimentai parodė, kad epochų skaičiaus padidinimas ir partijos dydžio keitimas turi teigiamą poveikį modelio tikslumui. Papildomo konvoliucinio sluoksnio pridėjimas taip pat pagerino modelio tikslumą.
